design/ceph/object/ceph-cosi-driver.md

```diff
@@ -0,0 +1,74 @@
+  verbs: ["get", "watch", "list", "delete", "update", "create"]
+- apiGroups: [""]
+  resources: ["secrets", "events"]
+  verbs: ["get", "delete", "update", "create"]
+```
+
+All the COSI CRDs are installed from <http://github.com/kubernetes-sigs/container-object-storage-interface-api>
+
+## Integration plan with Rook
+
+The aim to support alpha version of COSI in Rook v1.12 and depending on development of COSI, it will be extended to beta and GA versions in Rook Operaotr. There should be option in `Operator.yaml` to bring up the COSI controller and another option at object store level to bring up the ceph COSI driver.
+
+### How Rook can improve COSI driver reliability
+
+Rook can ensure that resources need for ceph cosi driver such as ceph object store is deployed and running in the cluster before creating requests for Bucket and its Access. Rook can also ensure that the COSI driver is deployed with the correct RBAC permissions, with the correct version of COSI controller. Rook should also prevent deletion of the ceph object Store if there are any buckets or bucket access in the cluster. Rook also need to provide secret containing the credentials for the ceph object store to the COSI driver. The secret need to updated if the ceph object store endpoint changes or credentials are changed. Rook need to bring up multiple drivers for  multiple ceph object stores in the cluster.
+
+### Coexistence of COSI and libbucket provisioner
+
+Currently the ceph object store provisioned via Object Bucket Claim (OBC). They both can co exist and can even use same backend bucket from ceph storage. No deployment/configuration changes are required to support both. The libbucket probvisioner is deprecated and eventually will be replaced by COSI when it becomes more and more stable. The CRDs used by both are different hence there is no conflicts between them.
+
+### Ceph COSI Driver Requirements
+
+- Ceph Object Store should be deployed and running in the cluster
+- The credentials/endpoint for the ceph object store should be available by creating ceph object store user with proper permissions
+- The COSI controller should be deployed in the cluster
+- Rook can able to manage multiple ceph cosi drivers
+- Rook should not modify cosi resource like Bucket, BucketAccess, BucketClaim, BucketAccessClass etc.
+
+### Rook Requirements
+
+- Rook must include all the RBAC permissions required by the COSI for the deployment.
+- Rook need to install all the COSI CRDs.
+- Rook need to dynamically create/update the secret containing the credentials of the ceph object store for ceph COSI driver.
+- User should not be required to deploy Rook differently when using COSI and OBC for ceph object store, expect the minimal changes in the `Operator.yaml` and `ObjectStore.yaml`.
+- When provisioning ceph COSI driver Rook must uniquely identify the driver/provisioner name so that multiple COSI drivers or multiple Rook instances within a (Kubernetes) cluster will not collide.
```

Comment:
Is it "driver" or "provisioner"? Let's pick one term and stick to it for the entire doc.

```suggestion
- When provisioning ceph COSI driver Rook must uniquely identify the driver/provisioner name so that multiple COSI drivers or multiple Rook instances within a (Kubernetes) cluster will not collide.
```

Reply:
missed this one, will address in next update